### PR TITLE
docs: add explanation, reference, and how to for bases

### PR DESCRIPTION
**Pull request by mr-cal**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

My task is to document how to use `base` and `build-base` for a kernel snap.  This required laying groundwork:

* import bases documentation from https://snapcraft.io/docs/base-snaps
    * This is a 5-year-old doc with minimal upkeep and wasn't a great starting point
* add [explanation](https://canonical-snapcraft--4900.com.readthedocs.build/en/4900/explanation/bases.html), [reference](https://canonical-snapcraft--4900.com.readthedocs.build/en/4900/reference/bases.html), and [how-to](https://canonical-snapcraft--4900.com.readthedocs.build/en/4900/howto/bases.html) pages for bases

(CRAFT-3083)

### CODE REVIEW

File: docs/reference/bases.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+--------
+
+The ``base`` keyword in a ``snapcraft.yaml`` file:
+
+* defines the feature set used by Snapcraft
+* the ``snapcraft.yaml`` schema
+* the environment where the snap is built if ``build-base`` is not defined
+* and which base snap is used at runtime
+
+``base`` must be defined except for base and snapd snaps.
```
Comment (mr-cal) [inline docs/reference/bases.rst:60]: This is a behavioral change from ST119.  Previously, `kernel` snaps also did not have to define `base`.  

We have to define a `base` in order to determine the feature set and snapcraft.yaml schema to use.

File: docs/explanation/bases.rst [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+While uncommon, developers can build their own base snap. They are responsible
+for maintenance and updates, in particular:
+
+* bases need to be built from stable packages
+* ABI compatibility cannot broken (ie. never replace symbols or libraries, and
+  be strict)
+* security updates must be proactive
+
+Base snaps can be either bootable or non-bootable. The former needs to include
+``systemd`` while the latter can be leaner.
```
Comment (cmatsuoka) [inline docs/explanation/bases.rst:66]: We can mention that bootable bases are only used in core systems (or is there another use case?)

**Comment by cmatsuoka** [comment]
Looks good! Added some notes about adding more information about building snaps with a bare bases, and keeping support to existing kernel snap projects (at least until they're phased out) to allow rebuilding of existing snaps.

File: docs/howto/bases.rst [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+See `Snapcraft and ESM`_ for details on support for deprecated bases.
+
+``core18``
+^^^^^^^^^^
+
+To build ``core18`` snaps, install snapcraft 7 from the *7.x* track:
+
+.. code-block:: shell
+
+   $ snap install snapcraft --channel 7.x
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/howto/bases.rst:43]: This? https://docs.ubuntu.com/styleguide/en#code-examples-in-documentation

File: docs/explanation/bases.rst [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+``base`` and ``build-base``
+---------------------------
+
+If ``build-base`` is defined , then the ``build-base`` will determine the
+environment where the snap is built.
+
+If ``build-base`` is not defined, the ``base`` will determine the
+environment where the snap is built.
+
+For example, ``base: core24`` will build a snap in a Ubuntu 24.04 environment
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/explanation/bases.rst:15]: I added information in both locations.

File: docs/reference/bases.rst (hunk 1 above)
Comment (mr-cal) [reply docs/reference/bases.rst:60]: Updated the reference to include baseless kernels.  I decided not to add a how-to for baseless kernels.

File: docs/explanation/bases.rst [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+A base snap is a special kind of snap that provides a run-time environment with
+a cardinal set of libraries that are common to most applications. They’re
+transparent to users, but they need to be considered and specified when building
+a snap.
+
+Mounting
+--------
+
+The base snap mounts itself as the root filesystem within a snap's runtime
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/explanation/bases.rst:34]: Updated

File: docs/explanation/bases.rst [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+environment. When an application runs, the base's library paths are searched
+directly after the paths for that snap.
+
+Choosing a base
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/explanation/bases.rst:38]: Updated

File: docs/explanation/bases.rst (hunk 2 above)
Comment (mr-cal) [reply docs/explanation/bases.rst:66]: Updated and added a link

File: docs/howto/bases.rst [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+Containers
+^^^^^^^^^^
+
+Isolated development environments allow using different versions of Snapcraft
+simultaneously.
+
+`Snapcraft rocks`_ are the recommended way to build snaps in a container:
+
+
+How to bootstrap a base snap
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/howto/bases.rst:99]: Added

File: docs/reference/bases.rst (hunk 1 above)
Comment (sergiusens) [reply docs/reference/bases.rst]: we could warn in this scenario and discourage its use in the documentation

**Comment by mr-cal** [comment]
Updated to only document the current way to build kernel snaps.

**Review by cmatsuoka — APPROVED** [review]
Thanks!

File: docs/howto/bases.rst [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+   snap install snapcraft snapcraft_8 --channel 8.x
+   snapcraft_8 pack
+
+Containers
+^^^^^^^^^^
+
+Isolated development environments allow using different versions of Snapcraft
+simultaneously.
+
+`Snapcraft rocks`_ are the recommended way to build snaps in a container:
```
Comment (mattculler) [inline docs/howto/bases.rst:96]:
```suggestion
`Snapcraft rocks`_ are the recommended way to build snaps in a container.
```

**Review by mattculler — APPROVED** [review]
Excellent, that was very informative and I learned a lot!

It seems to me that a lot of this content would be applicable to more than just snapcraft - should some of this documentation be upstreamed to the docs for the relevant library/libraries?

**Comment by mr-cal** [comment]
> Excellent, that was very informative and I learned a lot!
> 
> It seems to me that a lot of this content would be applicable to more than just snapcraft - should some of this documentation be upstreamed to the docs for the relevant library/libraries?

That's a good point.  I added some information in craft-application about bases and build-bases a few weeks ago.  As a future improvement, that would be very useful to expand upon in craft-application and pull that into the applications.

On that note, I'm not sure if craft-application documentation is configured to be re-used in the application documentation.

**Comment by mr-cal** [comment]
I did pull in some of your `craft-parts` documentation on partitions here.
I'll defer the `craft-application` documentation updates/inclusions since it's friday afternoon!

**Comment by mr-cal** [comment]
Spread tests failures unrelated, see #4910

**Review by mattculler — APPROVED** [review]

File: docs/explanation/bases.rst [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+Mounting
+--------
+
+For strictly confined snaps, the base snap mounts itself as the root filesystem
+within a snap's runtime environment. When an application runs, the base's
+library paths are searched directly after the paths for that snap.
+
+For classic confined snaps, the base snap is not mounted as the root
+filesystem. The base snap is mounted like as ``/snap/<base>/``, so the snap can
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply docs/explanation/bases.rst:42]: I need to find a grammar checker for nvim!

**Pull request merged**